canjs-devtools-helpers.mjs

```diff
@@ -1,231 +1,240 @@
+		};
+
+		const frameChangeHandler = () => {
+			// remove old `refreshDataForThisUrl` timeouts
+			teardown();
+
+			// run function for all new frames
+			runDevtoolsFunctionForAllUrls();
+		};
+
+		const runDevtoolsFunctionForAllUrls = () => {
+			const frameURLs = Object.keys(helpers.registeredFrames);
+
+			if (frameURLs.length) {
+				frameURLs.forEach(url => {
+					runDevtoolsFunctionForUrl(url);
+				});
+			}
+
+			// when a frame is added or removed
+			// remove the old refresh handlers
+			// and re-create them
+			this.onFrameChange(frameChangeHandler);
+		};
+
+		runDevtoolsFunctionForAllUrls();
+
+		return teardown;
+	},
+
+	getSafeKey(key, prependStr) {
+		const parts = key.split(".");
+		let last = "";
+
+		const safeParts = parts.reduce((newParts, key) => {
+			last = last ? `${last}.${key}` : `${prependStr}.${key}`;
+			newParts.push(last);
+			return newParts;
+		}, []);
+
+		if (parts.length > 1) {
+			return `(${safeParts.join(" && ")})`;
+		} else {
+			return safeParts.join(" && ");
+		}
+	},
+
+	getObservationExpression(expression) {
+		return expression.replace(
+			/(^|\s|=|>|<|!|\/|%|\+|-|\*|&|\(|\)|~|\?|,|\[|\])([A-Za-z_:.]*)/g,
+			(match, delimiter, key) => {
+				return `${delimiter}${key ? this.getSafeKey(key, "vm") : ""}`;
+			}
+		);
+	},
+
+	getDisplayExpression(expression) {
+		return expression.replace(
+			/(^|\s|=|>|<|!|\/|%|\+|-|\*|&|\(|\)|~|\?|,|\[|\])([A-Za-z_])/g,
+			(match, delimiter, prop) => {
+				return `${delimiter}\$\{vmName\}.${prop}`;
+			}
+		);
+	},
+
+	isBooleanExpression(expression) {
+		return /[!=<>]/.test(expression);
+	},
+
+	getBreakpointEvalString({
+		expression,
+		enabled = true,
+		observationExpression = this.getObservationExpression(expression),
+		selectedComponentStatement = "window.__CANJS_DEVTOOLS__.$0",
+		displayExpression = this.getDisplayExpression(expression),
+		pathStatement = "window.__CANJS_DEVTOOLS__.pathOf$0",
+		debuggerStatement = "debugger" // overwritable for testing
+	}) {
+		const isBooleanExpression = this.isBooleanExpression(observationExpression);
+
+		return `(function() {
+				const Observation = window.__CANJS_DEVTOOLS__.canObservation;
+				const queues = window.__CANJS_DEVTOOLS__.canQueues;
+				const selectedComponent = ${selectedComponentStatement};
+
+				if (!selectedComponent) {
+						return { error: "Please select a component in order to create a mutation breakpoint for its ViewModel" };
+				}
+
+				const vm = selectedComponent[Symbol.for('can.viewModel')];
+				const vmName = window.__CANJS_DEVTOOLS__.canReflect.getName(vm);
+				let oldValue = ${observationExpression};
+
+				const observation = new Observation(() => {
+						return ${observationExpression};
+				});
+				const origDependencyChange = observation.dependencyChange;
+
+				observation.dependencyChange = function() {
+						const newValue = ${observationExpression};
+						${
+							isBooleanExpression
+								? `if (newValue == true && oldValue != true) {
```

Comment:
Is this indented weird?

Reply:
it looks weird. I just ran prettier, sometimes it does weird stuff on string literals.